File: android/src/main/kotlin/com/eopeter/fluttermapboxnavigation/boundings/location/application/Codec.kt
package com.eopeter.fluttermapboxnavigation.boundings.location.application

import com.eopeter.fluttermapboxnavigation.core.*
import io.flutter.plugin.common.StandardMessageCodec
import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer

@Suppress("UNCHECKED_CAST")
object LocationApiCodec: StandardMessageCodec() {
    override fun readValueOfType(type: Byte, buffer: ByteBuffer): Any? {
        return when (type) {
            128.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    LocationComponentSettings.fromList(it)
                }
            }
            else -> super.readValueOfType(type, buffer)
        }
    }
    override fun writeValue(stream: ByteArrayOutputStream, value: Any?) {
        when (value) {
            is LocationComponentSettings -> {
                stream.write(128)
                writeValue(stream, value.toList())
            }
            else -> super.writeValue(stream, value)
        }
    }
}

File: lib/src/embedded/boundings/location/application/codec.dart
part of mapbox_navigation_flutter;

class LocationAPICodec extends StandardMessageCodec {
  const LocationAPICodec() : super();

  @override
  void writeValue(WriteBuffer buffer, Object? value) {
    if (value is LocationComponentSettings) {
      buffer.putUint8(128);
      writeValue(buffer, value.encode());
    } else {
      super.writeValue(buffer, value);
    }
  }

  @override
  Object? readValueOfType(int type, ReadBuffer buffer) {
    if (type == 128) {
      return LocationComponentSettings.decode(readValue(buffer)!);
    } else {
      return super.readValueOfType(type, buffer);
    }
  }
}
